xformers/csrc/attention/hip_fmha/attention_forward_decoder.cpp

```diff
@@ -0,0 +1,333 @@
+/*
```

Comment:
So far the decoding perf looks like this -- 

```
[------------------------------------------------------------------- attn_decodingfw -------------------------------------------------------------------]
                                             |  pytorch  |  optimized[ck]  |  optimized[ck-decoder]  |  optimized[ck_splitK]  |  optimized[triton_splitK]
1 threads: ----------------------------------------------------------------------------------------------------------------------------------------------
      B=256 Mq=1 Mkv=256 Hq=16 Hkv=1 K=128   |   1474.1  |       169.8     |          175.7          |         234.6          |            71.7          
      B=256 Mq=1 Mkv=256 Hq=16 Hkv=2 K=128   |   2407.5  |       294.6     |          188.0          |         240.8          |           112.4          
      B=128 Mq=1 Mkv=512 Hq=16 Hkv=1 K=128   |   1576.4  |       185.1     |          176.2          |         255.1          |           104.1          
      B=128 Mq=1 Mkv=512 Hq=16 Hkv=2 K=128   |   2456.4  |       291.6     |          173.4          |         228.0          |            75.8          
      B=64 Mq=1 Mkv=1024 Hq=16 Hkv=1 K=128   |   1479.6  |       175.7     |          158.2          |         209.2          |            71.4          
      B=64 Mq=1 Mkv=1024 Hq=16 Hkv=2 K=128   |   2486.5  |       347.7     |          185.6          |         274.0          |           107.8          
      B=32 Mq=1 Mkv=2048 Hq=16 Hkv=1 K=128   |   1405.4  |       174.8     |          149.0          |         210.2          |            67.4          
      B=32 Mq=1 Mkv=2048 Hq=16 Hkv=2 K=128   |   2259.1  |       276.6     |          173.8          |         248.0          |            77.4          
      B=16 Mq=1 Mkv=4096 Hq=16 Hkv=1 K=128   |   1395.6  |       211.0     |          145.4          |         230.4          |            70.3          
      B=16 Mq=1 Mkv=4096 Hq=16 Hkv=2 K=128   |   2236.9  |       413.0     |          172.0          |         400.2          |            81.4          
      B=8 Mq=1 Mkv=8192 Hq=16 Hkv=1 K=128    |   1350.1  |       354.3     |          221.5          |         588.4          |            71.9          
      B=8 Mq=1 Mkv=8192 Hq=16 Hkv=2 K=128    |   2192.8  |       734.6     |          232.2          |         603.2          |            81.8          
      B=4 Mq=1 Mkv=16384 Hq=16 Hkv=1 K=128   |   1674.7  |       656.2     |            N/A*          |         228.3          |            86.3          
      B=4 Mq=1 Mkv=16384 Hq=16 Hkv=2 K=128   |   2686.9  |      1432.5     |            N/A*          |         246.5          |            75.1          
      B=2 Mq=1 Mkv=32768 Hq=16 Hkv=1 K=128   |   2206.1  |      1289.7     |            N/A*          |         204.2          |            72.2          
      B=2 Mq=1 Mkv=32768 Hq=16 Hkv=2 K=128   |   3592.0  |      2702.8     |            N/A*          |         221.9          |            81.7          
      B=1 Mq=1 Mkv=65536 Hq=16 Hkv=1 K=128   |    635.6  |      2557.7     |            N/A*          |         201.2          |            92.3          
      B=1 Mq=1 Mkv=65536 Hq=16 Hkv=2 K=128   |   1207.9  |      5340.1     |            N/A*          |         225.3          |            95.4          
      B=1 Mq=1 Mkv=131072 Hq=16 Hkv=1 K=128  |   1181.4  |      5107.4     |            N/A*          |         314.7          |           139.1          
      B=1 Mq=1 Mkv=131072 Hq=16 Hkv=2 K=128  |   2250.4  |     10632.0     |            N/A*          |         359.9          |           143.0

* Max Mkv = 8192 for ck decoder because we run out of shared memory for larger sequences          
```

So it looks like at this point each implementation in (ck, ck_decoder, ck_splitk) has its best case for decoding, and we need to do some perf tuning